File: 0082-remove-duplicates-from-sorted-list-ii/0082-remove-duplicates-from-sorted-list-ii.py
# Definition for singly-linked list.
# class ListNode:
#   def __init__(self, val=0, next=None):
#     self.val = val
#     self.next = next
class Solution:
  def deleteDuplicates(self, head: Optional[ListNode]) -> Optional[ListNode]:
    if head is None:
      return head
    initVal = head.val
    prev = ListNode()
    prev.next = curr = head
    while curr:
      if curr.next and curr.next.val != curr.val:
        prev = prev.next
        curr = curr.next
      elif curr.next and curr.next.val == curr.val:
        while curr.next and curr.next.val == curr.val:
          curr = curr.next
        prev.next = curr.next
        if curr.val == initVal:
          head = prev.next
          if head:
            initVal = head.val
        if curr.next is None:
          return head
        else:
          curr = prev.next
      else:
        return head
        

File: 0082-remove-duplicates-from-sorted-list-ii/0082-remove-duplicates-from-sorted-list-ii.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func deleteDuplicates(head *ListNode) *ListNode {
  if head == nil {
    return head
  }
  initVal, prev, curr := head.Val, &ListNode{}, head
  prev.Next = head
  for curr != nil {
    if curr.Next != nil && curr.Next.Val != curr.Val {
      prev, curr = prev.Next, curr.Next
    } else if curr.Next != nil && curr.Next.Val == curr.Val {
      for curr.Next != nil && curr.Next.Val == curr.Val {
        curr = curr.Next
      }
      prev.Next = curr.Next
      if curr.Val == initVal {
        head = prev.Next
        if head != nil {
          initVal = head.Val
        }
      }
      if curr.Next == nil {
        break
      } else {
        curr = prev.Next
      }
    } else {
      break
    }
  }
  return head
}